community/kernel/src/main/java/org/neo4j/helpers/SortedLongArrayUtil.java

```diff
@@ -0,0 +1,251 @@
+     */
+    public static long[] intersect( long[] left, long[] right )
+    {
+        if ( left == null || right == null )
+        {
+            return EMPTY;
+        }
+
+        long uniqueCounts = countUnique( left, right );
+        if ( uniqueCounts == 0 ) // complete intersection
+        {
+            return right;
+        }
+        if ( right( uniqueCounts ) == right.length || left( uniqueCounts ) == left.length ) // non-intersecting
+        {
+            return EMPTY;
+        }
+
+        long[] intersect = new long[left.length - left( uniqueCounts )];
+
+        int cursor = 0;
+        int l = 0;
+        int r = 0;
+        while ( l < left.length && r < right.length )
+        {
+            if ( left[l] == right[r] )
+            {
+                intersect[cursor++] = left[l];
+                l++;
+                r++;
+            }
+            else if ( left[l] < right[r] )
+            {
+                l++;
+            }
+            else
+            {
+                r++;
+            }
+        }
+
+        assert cursor == intersect.length;
+        return intersect;
+    }
+
+    /**
+     * Compute the symmetric difference (set XOR basically) of two sorted long array sets.
+     * @param left a sorted array set
+     * @param right another sorted array set
+     * @return the union, which is NOT sorted
+     */
+    public static long[] symmetricDifference( long[] left, long[] right )
+    {
+        if ( left == null || right == null )
+        {
+            return left == null ? right : left;
+        }
+
+        long uniqueCounts = countUnique( left, right );
+        if ( uniqueCounts == 0 ) // complete intersection
+        {
+            return EMPTY;
+        }
+
+        long[] difference = new long[left( uniqueCounts ) + right( uniqueCounts )];
+
+        int cursor = 0;
+        int l = 0;
+        int r = 0;
+        while ( l < left.length && r < right.length )
+        {
+            if ( left[l] == right[r] )
+            {
+                l++;
+                r++;
+            }
+            else if ( left[l] < right[r] )
+            {
+                difference[cursor++] = left[l];
+                l++;
+            }
+            else
+            {
+                difference[cursor++] = right[r];
+                r++;
+            }
+        }
+        while ( l < left.length )
+        {
+            difference[cursor++] = left[l];
+            l++;
+        }
+        while ( r < right.length )
+        {
+            difference[cursor++] = right[r];
+            r++;
+        }
+
+        assert cursor == difference.length;
+        return difference;
+    }
+
+    /**
+     * Compute the number of unique values in two sorted long array sets
+     * @param left a sorted array set
+     * @param right another sorted array set
+     * @return int pair packed into long
+     */
+    public static long countUnique( long[] left, long[] right )
```

Comment:
This method is not used outside of the implementation of the methods above (and in tests), since it has a weird return type, I vote that it be made a private implementation detail rather than part of the API (or package-private in order to allow testing).

If the above methods no longer depend on this method when they return sorted output instead (which might become the case), I vote that we remove this method.